### PR TITLE
#MultiColorLayout does formatting, parsing, coloring and managing human-readable color names and should be refactored.

### PR DESCRIPTION
**Pull request by piddubnyi**
Solving #59 
Extracted Color Manager so now managing human-readable color names is done in a separate class.

### CODE REVIEW

**Comment by dmarkov** [comment]
@piddubnyi Thanks for the pull request, let me find a reviewer..

**Comment by dmarkov** [comment]
@mkordas could you please review this pull request

**Comment by mkordas** [comment]
@piddubnyi I'm on it

File: src/main/java/com/jcabi/log/ColorManager.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+package com.jcabi.log;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Color manager to store color data.
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 1.0
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/ColorManager.java]: @piddubnyi should be `@since 0.18`

File: src/main/java/com/jcabi/log/ColorManager.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ */
+public class ColorManager {
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/ColorManager.java]: @piddubnyi please read http://www.yegor256.com/2015/03/09/objects-end-with-er.html and watch https://www.youtube.com/watch?v=WpP4rIhh5e4

File: src/main/java/com/jcabi/log/ColorManager.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * Colors with names.
+     */
+    private final transient ConcurrentMap<String, String> colors =
+        ColorManager.colorMap();
+
+    /**
+     * Add color to color map.
+     * @param key Key to add.
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/ColorManager.java]: @piddubnyi let's be consistent and not add dots at the end of at-tags as you did a few lines below

File: src/main/java/com/jcabi/log/ColorManager.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        return ansi;
+    }
+
+    /**
+     * Color map.
+     * @return Map of colors
+     */
+    private static ConcurrentMap<String, String> colorMap() {
+        final ConcurrentMap<String, String> map =
+            new ConcurrentHashMap<String, String>();
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/ColorManager.java]: @piddubnyi We're not modifying that map after creation, I think we should use `ImmutableMap` instead

File: src/main/java/com/jcabi/log/ColorManager.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+ * @version $Id$
+ * @since 1.0
+ */
+public class ColorManager {
+
+    /**
+     * Colors with names.
+     */
+    private final transient ConcurrentMap<String, String> colors =
+        ColorManager.colorMap();
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/ColorManager.java]: @piddubnyi please move that field initialization to constructor

File: src/main/java/com/jcabi/log/MulticolorLayout.java [hunk 6]
```diff
@@ -75,7 +75,7 @@
  *  &lt;groupId&gt;com.jcabi&lt;/groupId&gt;
  *  &lt;artifactId&gt;jcabi-log&lt;/artifactId&gt;
  * &lt;/dependency&gt;</pre>
- *
+ * @todo #59:30min Continue refactoring of the class started in #51.
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/MulticolorLayout.java]: @piddubnyi can you provide more details here? Like what was already done and what is still left?

File: src/main/java/com/jcabi/log/MulticolorLayout.java [hunk 7]
```diff
@@ -137,6 +131,11 @@
      */
     private transient String base;
 
+    /**
+     * Helper class to store color data.
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/MulticolorLayout.java]: @piddubnyi
> Helper class

Please look at http://blogs.msdn.com/b/nickmalik/archive/2005/09/06/461404.aspx and http://www.smart421.com/application-development/why-helper-singletons-and-utility-classes-are-mostly-bad-2/

**Comment by mkordas** [comment]
@piddubnyi please see a few comments above

**Comment by piddubnyi** [comment]
@mkordas I dont see an easy way to decouple logic of this class, that's why I did first and easiest thing: I moved logic for managing human readable color names to another class, and Introduced a puzzle to continue on refactoring.
What is your opinion, how can we solve initial issue #59 ?

**Comment by mkordas** [comment]
@piddubnyi that approach of extracting code about colors is fine, but again, please read http://www.yegor256.com/2015/03/09/objects-end-with-er.html and watch https://www.youtube.com/watch?v=WpP4rIhh5e4

**Comment by mkordas** [comment]
@piddubnyi moreover, please try to fix the build

File: src/main/java/com/jcabi/log/ColorManager.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 0.18
+ */
+public class ColorManager {
+
+    /**
+     * Colors with names.
+     */
+    private transient ConcurrentMap<String, String> colors;
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/ColorManager.java]: @piddubnyi shuld be `final`

**Comment by mkordas** [comment]
@piddubnyi introducing a puzzle is good approach to partially solve issue for now, but you need to put more description there

**Comment by mkordas** [comment]
@piddubnyi actually build is failing because of the following:
```
[ERROR] \src\main\java\com\jcabi\log\ColorManager.java[46-46]: Private field 'colors' could be made final; it is only initialized in the declaration or constructor. (ImmutableField)
```

**Comment by piddubnyi** [comment]
@mkordas Please check.

File: src/main/java/com/jcabi/log/Colors.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jcabi.log;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Color manager to store color data.
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/Colors.java]: @piddubnyi please adjust Javadoc

File: src/main/java/com/jcabi/log/MulticolorLayout.java [hunk 10]
```diff
@@ -75,7 +75,8 @@
  *  &lt;groupId&gt;com.jcabi&lt;/groupId&gt;
  *  &lt;artifactId&gt;jcabi-log&lt;/artifactId&gt;
  * &lt;/dependency&gt;</pre>
- *
+ * @todo #59:30min This class is still handling multiple responsibilities like
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/MulticolorLayout.java]: @piddubnyi `@todo` tag should be placed at the end of Javadoc

File: src/main/java/com/jcabi/log/MulticolorLayout.java (hunk 7 above)
Comment (mkordas) [inline src/main/java/com/jcabi/log/MulticolorLayout.java]: @piddubnyi please replace 'helper class' with something different

File: src/main/java/com/jcabi/log/MulticolorLayout.java [hunk 11]
```diff
@@ -137,6 +132,11 @@
      */
     private transient String base;
 
+    /**
+     * Helper class to store color data.
+     */
+    private final transient Colors colrs = new Colors();
```
Comment (mkordas) [inline src/main/java/com/jcabi/log/MulticolorLayout.java]: @piddubnyi typo: should be `colors`

**Comment by mkordas** [comment]
@piddubnyi please see a few comments above

**Comment by piddubnyi** [comment]
@caarlos0 Fixed.

**Comment by caarlos0** [comment]
@piddubnyi ?

**Comment by piddubnyi** [comment]
@caarlos0 sorry, wrong name. I meant @mkordas

**Comment by mkordas** [comment]
@piddubnyi good job!

**Comment by mkordas** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by mkordas** [comment]
@rultor deploy

**Comment by mkordas** [comment]
@yegor256 ping, it's waiting for merge 4 days already

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor deploy

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by mkordas** [comment]
@rultor deploy

**Comment by yegor256** [comment]
@rultor deploy

**Comment by dmarkov** [comment]
@mkordas paid, thanks, added **38 mins** to your account, payment `56b31b084f694f4168000195`, 119 hours and 18 mins was spent; extra minutes for review comments (c=23); +38 added to your rating, current score is: [+5002](http://www.netbout.com/b/36080?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy

**Pull request merged**